hidparser/src/report_descriptor_parser.rs

```diff
@@ -79,15 +80,15 @@ struct ReportData {
 impl ReportData {
   fn new(
     item_data: &[u8],
-    global_state: &GlobalItemStateTable,
-    local_state: &LocalItemStateTable,
-    active_collections: &[ReportCollection],
+    global_state: GlobalItemStateTable,
+    local_state: LocalItemStateTable,
+    active_collections: Vec<ReportCollection>,

```

Comment:
You should always do `$[]` for function variables as rust will automatically coerce types to a slice, where as this locks the input to only a vector type.

Reply:
The true best would be to use a generic:

```rust
fn new<I: IntoIterator<Item = ReportCollection>>(..., active_collections: I) { /* ... */ }
```

Reply:
I agree that generally it is preferable to use &[] or a generic in functions, but since the `new` function take total ownership of the data, it is preferable to pass the own type in argument, so we don't have to clone data. With a slice reference, the only way to end up with a Vec is to copy its content in a new Vec. The same reasoning applies to using a generic. If we want to use a generic it would be best to put the struct generic too.